tests/unit/states/test_vmc_dhcp_profiles.py

```diff
@@ -0,0 +1,576 @@
+"""
+    Unit tests for vmc_dhcp_profiles state module
+"""
+from unittest.mock import create_autospec
+from unittest.mock import patch
+
+import pytest
+import saltext.vmware.modules.vmc_dhcp_profiles as vmc_dhcp_profiles_exec
+import saltext.vmware.states.vmc_dhcp_profiles as vmc_dhcp_profiles
+
+
+@pytest.fixture
+def configure_loader_modules():
+    return {vmc_dhcp_profiles: {}}
+
+
+@pytest.fixture
+def get_mocked_data():
```

Comment:
What is your reasoning behind this being a fixture, a function, 3 functions or 3 module-level variables?

Reply:
3 module-level variables

Reply:
this is confusing, can we discuss over call?

Reply:
@shaikres IIUC what @dhiltonp is suggesting is that:

```
@pytest.fixture
def ok_response():
    response = {...}
    return response
    
@pytest.fixture
def error_response():
    response = {'error': ...}
    return response

...
```

Would be more correct... and he's correct. Though it would be even more correct to shift the *mocks* into the fixtures. The point of fixtures is to allow the de-duplication of code by moving mocks or other common setup out of test functions. It also allows tests to be very narrow in scope. So instead of

```
def test_present_state_when_error_from_get_by_id(get_mocked_data):
    mocked_hostname, mocked_ok_response, mocked_error_response = get_mocked_data
    mock_get_by_id = create_autospec(
        vmc_dhcp_profiles_exec.get_by_id, return_value=mocked_error_response
    )

    with patch.dict(vmc_dhcp_profiles.__salt__, {"vmc_dhcp_profiles.get_by_id": mock_get_by_id}):
        result = vmc_dhcp_profiles.present(
            name="test_present",
            hostname=mocked_hostname,
            refresh_key="refresh_key",
            authorization_host="authorization_host",
            org_id="org_id",
            sddc_id="sddc_id",
            type="relay",
            dhcp_profile_id=mocked_ok_response["id"],
        )
```

You would have

```
@pytest.fixture
def mocked_error_response():
    error_response = {'error': ...}
    mock_get_by_id = create_autospec(
        vmc_dhcp_profiles_exec.get_by_id, return_value=error_response
    )
    with patch.dict(vmc_dhcp_profiles.__salt__, {"vmc_dhcp_profiles.get_by_id": mock_get_by_id}):
        yield
        

def test_present_state_when_error_from_get_by_id(mocked_error_response):
    result = vmc_dhcp_profiles.present(
            name="test_present",
            hostname="fnord",   # or some other string. Since you're mocking `get_by_id`, and this is not in any assertion, there's no reason make it look "real". And by explicitly making it something not real, you're indicating to future readers that the value of hostname is irrelevant to this test.
            refresh_key="refresh_key",
            authorization_host="authorization_host",
            org_id="org_id",
            sddc_id="sddc_id",
            type="relay",
            dhcp_profile_id="same fnord thing here"],
    )

    assert result is not None
    assert result["changes"] == {}
    assert (
        result["comment"]
        == "The credentials were incorrect or the account specified has been locked."
    )
    assert not result["result"]
```

If this is unclear, I can review it on one of the [Test Clinics on Twitch](https://www.twitch.tv/saltstackinc/schedule).

Reply:
Thanks @waynew for clarifying this point.
I updated the testcases accordingly and pushed it to the PR: [https://github.com/saltstack/salt-ext-modules-vmware/pull/130](url)